de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/SarosSessionManager.java

```diff
@@ -532,66 +588,43 @@ public void addResourcesToSession(
             return;
         }
 
-        List<IProject> projectsToShare = new ArrayList<IProject>();
-
-        for (Entry<IProject, List<IResource>> mapEntry : projectResourcesMapping
-            .entrySet()) {
-
-            final IProject project = mapEntry.getKey();
-            final List<IResource> resourcesList = mapEntry.getValue();
-
-            // side effect: non shared projects are always partial -.-
-            if (!currentSession.isCompletelyShared(project)) {
-                String projectID = currentSession.getProjectID(project);
-
-                if (projectID == null)
-                    projectID = String.valueOf(SESSION_ID_GENERATOR
-                        .nextInt(Integer.MAX_VALUE));
-
-                currentSession.addSharedResources(project, projectID,
-                    resourcesList);
-
-                projectsToShare.add(project);
-            }
-        }
+        List<IProject> projectsToShare = nextProjectNegotiation
+            .getProjectsToShare(currentSession, SESSION_ID_GENERATOR);
 
         if (projectsToShare.isEmpty()) {
             log.warn("skipping project negotiation because no new projects were added to the current session");
             return;
         }
 
         INegotiationHandler handler = negotiationHandler;
-
         if (handler == null) {
             log.warn("could not start a project negotiation because no handler is installed");
             return;
         }
 
         List<AbstractOutgoingProjectNegotiation> negotiations = new ArrayList<AbstractOutgoingProjectNegotiation>();
 
-        synchronized (this) {
-            if (!startStopSessionLock.tryLock()) {
-                log.warn("could not start a project negotiation because the current session is about to stop");
-                return;
-            }
+        if (!startStopSessionLock.tryLock()) {
```

Comment:
Ok now I remember my logic. I used synchronized as second lock to avoid a possible situation where adding project and receiving a project concurrently would result in one of the operation being discarded. However this is currently not possible due to the architecture (only the host can add projects and only client can receive projects). We still have a patch from Denis that try to reintroduce this behavior for the Saros server. So I do not think moving the sychronized block is the best option here as you increase the size of the code that is being locked.